vendor/k8s.io/kubernetes/pkg/kubelet/dockershim/helpers.go

```diff
@@ -322,13 +322,12 @@ func ensureSandboxImageExists(client libdocker.Interface, image string) error {
 		return fmt.Errorf("failed to inspect sandbox image %q: %v", image, err)
 	}
 
-	repoToPull, _, _, err := parsers.ParseImageName(image)
-	if err != nil {
+	if _, _, _, err := parsers.ParseImageName(image); err != nil {
 		return err
 	}
 
 	keyring := credentialprovider.NewDockerKeyring()
-	creds, withCredentials := keyring.Lookup(repoToPull)
+	creds, withCredentials := keyring.Lookup(image)
```

Comment:
it would be good to know what happens when this gets an unqualified image name.

is it using the first credential in the keyring, the docker.io credential, etc.?

Reply:
also need to know what credentials are used in this case by docker daemon.

Reply:
looks like working back in the code, it will return all credentials registered w/ the provider.

Reply:
> also need to know what credentials are used in this case by docker daemon.

I believe that if the image reference is unqualified then you'll get no credentials at all; I tried to summarise what I found here: [https://bugzilla.redhat.com/show_bug.cgi?id=1583500#c39](https://bugzilla.redhat.com/show_bug.cgi?id=1583500#c39)

Reply:
I should add that my statement about "you'll get no credentials" is with this patch applied. Without this patch unqualified images become "docker.io" and in the keyring we'll match against:

	// Use credentials for the default registry if provided, and appropriate
	if isDefaultRegistryMatch(image) {
		if auth, ok := dk.creds[defaultRegistryHost]; ok {
			return auth, true
		}
	}

Reply:
Actually, AFAICT `isDefaultRegistryMatch` (https://github.com/openshift/origin/blob/2fa80426cf1fd0c8e67ae46692bad380ca25910a/vendor/k8s.io/kubernetes/pkg/credentialprovider/keyring.go#L137 ) is `true` for unqualified images (it implements the `docker.io` normalization internally).

So, this patch changes the string that is passed to `Lookup` but not the outcome (`docker.io/`_image_ credentials are returned).

The user-visible impact is unchanged, users should always specify fully-qualified name if the first registry in the search order requires authentication.

---

Another aspect of this is that the keyrings are hostname-scoped; there is no semantically right place to look for answers to `keyring.Lookup("nohostname")`, at least without knowing what the node’s list of registries to search is.

`projectatomic/docker`’s CLI handles `docker pull nohostname` by reading the whole config, and sending _all_ hostname→credentials pairs to the daemon in the pull API call — but that’s a `projectatomic/docker` extension which is not understood `moby/moby` engine API servers; IIRC they treat such calls as if no credentials were supplied at all.